airflow/models.py

```diff
@@ -2207,7 +2237,7 @@ def __init__(
             on_success_callback=None,
             on_retry_callback=None,
             trigger_rule=TriggerRule.ALL_SUCCESS,
-            resources=None,
+            resources=Resources(),
```

Comment:
the Resources object will be shared across invocations here

Reply:
Yes that's a potential breaking change. I couldn't find where a task's `resources` was actually being modified anywhere after construction and I didn't quite understand its usage for when that would actually happen. I only saw this being accessed in `cgroup_task_runner`. Did I miss a use case here? Additionally, I added a change to prevent modifying an existing Resource object as well.

Reply:
I _think_ that this change is okay, especially with you having made the object's properties read-only.  It might be worth a comment somewhere saying "shared by design" with a small bit of rationale

Reply:
I’m gonna vote for removal. Even if read only now, it may change in the future. Also since it’s usually a bad pattern, I would call it bad style/confusing.

Reply:
Just to make sure we're on the same page here this was a pretty big performance sink with the 4x configuration.getInt requests for every new operator.

Width | Tasks | New Resource Object | Static Resources Object
-- | -- | -- | --
12800 | 51205 | 4.85 | 2.7
51200 | 204805 | 18 | 10.3

I'm ok with pulling this commit out:
[Share default Resources object when no Resources specified 6da941e](https://github.com/apache/incubator-airflow/pull/3116/commits/6da941e808a2e1e858e1c20ecf0df70b3bd3d932)

That said, users aren't completely forced into this performance sink, just a bit clunky to have ask users to manually share a default Resources object if they care about performance of very large dags. I can add a comment to advise this if they are.

As an alternative, what do you think about this diff in operators_resources.py`? at least then we don't have to load the configuration 4x with every new operator (interpolating/checking env vars, checking `airflow.cfg`, etc ...)
```
+++ b/airflow/utils/operator_resources.py
@@ -88,7 +88,9 @@ class GpuResource(Resource):
 class Resources(object):
-    def __init__(self, cpus=None, ram=None, disk=None, gpus=None):
-        if cpus is None:
-            cpus = configuration.getint('operators', 'default_cpus')
-        if ram is None:
-            ram = configuration.getint('operators', 'default_ram')
-        if disk is None:
-            disk = configuration.getint('operators', 'default_disk')
-        if gpus is None:
-            gpus = configuration.getint('operators', 'default_gpus')
+    def __init__(self,
+                 cpus=configuration.getint('operators', 'default_cpus'),
+                 ram=configuration.getint('operators', 'default_ram'),
+                 disk=configuration.getint('operators', 'default_disk'),
+                 gpus=configuration.getint('operators', 'default_gpus')
+                 ):

```

Thanks for reviewing! 
Let me know what you would like me to do! 😃

Reply:
Yeah, I think this change sounds like a good compromise.

(This now makes me wonder just how slow ConfigParser is and if it's causing any other bottlenecks in Airflow.... 😟 )